README.md
# NAME

HTTP::GHTTP - (DEPRECATED) Perl interface to the gnome ghttp library

# SYNOPSIS

    use strict;
    use warnings;
    use HTTP::GHTTP;

    my $r = HTTP::GHTTP->new();
    $r->set_uri("http://axkit.org/");
    $r->process_request;
    print $r->get_body;

# DESCRIPTION

The GNOME [libghttp](http://ftp.gnome.org/pub/gnome/sources/libghttp) project
has been dead since 2002.  Thus, this library is deprecated.

**Do not use this module!**

This is a fairly low level interface to [libghttp](http://ftp.gnome.org/pub/gnome/sources/libghttp),
which allows you to process HTTP requests to HTTP servers. There also
exists a slightly higher level interface - a simple get() function
which takes a URI as a parameter. This is not exported by default, you
have to ask for it explicitly.

# FUNCTIONS

The [HTTP::GHTTP](https://metacpan.org/pod/HTTP::GHTTP) module makes the following standalone functions available.

## get

    get($uri, [%headers])

This does everything automatically for you, retrieving the body at
the remote URI. Optionally pass in headers.

# METHODS

The [HTTP::GHTTP](https://metacpan.org/pod/HTTP::GHTTP) module makes the following methods available.

## new

    my $r = HTTP::GHTTP->new("http://axkit.com/", Connection => "close");

Constructor function - creates a new GHTTP object. If supplied a URI
it will automatically call set\_uri for you. If you also supply a list
of key/value pairs it will set those as headers:

## current\_status

    $r->current_status()

This is only useful in async mode. It returns 3 values: The current
processing stage (0 = none, 1 = request, 2 = response headers,
3 = response), the number of bytes read, and the number of bytes total.

## get\_body

    my $rresponse = $r->get_body();

This gets the body of the response

## get\_error

    my $error = $r->get_error();

If the response failed for some reason, this returns a textual error.

## get\_header

    my $value = $r->get_header($header);

This gets the value of an incoming HTTP response header.

## get\_headers

    my @headers = $r->get_headers;
    print join("\n",
      map { "$_: " . $r->get_header($_) } @headers), "\n\n";

Returns a list of all the response header names in the order they
came back. This method is only available in [libghttp](http://ftp.gnome.org/pub/gnome/sources/libghttp) 1.08 and later -
`perl Makefile.PL` should have reported whether it found it or not.

## get\_socket

    my $sock_h = $r->get_socket();

Returns an [IO::Handle](https://metacpan.org/pod/IO::Handle) object that is the currently in progress socket.
Useful only when doing async downloads. There appears to be some corruption
when using the socket to retrieve file contents on more recent [libghttp](http://ftp.gnome.org/pub/gnome/sources/libghttp).

## get\_status

    $r->get_status()

This returns 2 values, a status code (numeric) and a status reason
phrase. A simple example of the return values would be (200, "OK").

## prepare

    $r->prepare()

This is a low level interface useful only when doing async downloads.
See ["ASYNC OPERATION"](#async-operation).

## process

    $r->process()

This is a low level interface useful only when doing async downloads.
See ["ASYNC OPERATION"](#async-operation).

process returns `undef` for error, 1 for "in progress", and zero for
"complete".

## process\_request

    $r->process_request()

This sends the actual request to the server

## set\_async

    $r->set_async()

This turns async mode on. There is no corresponding unset function.

## set\_authinfo

    $r->set_authinfo($user, $password)

This sets an outgoing username and password for simple HTTP
authentication.

## set\_body

    $r->set_body($body);

This sets the body of a request, useful in `POST` and some of the DAV
request types.

## set\_chunksize

    $r->set_chunksize($bytes)

Sets the download (and upload) chunk size in bytes for use in async
mode. This may be a useful value to set for slow modems, or perhaps
for a download progress bar, or just to allow periodic writes.

## set\_header

    $r->set_header($header_name, $header_value);
    $r->set_header('Connection', 'Close');

This sets an outgoing HTTP request header.

## set\_proxy

    $r->set_proxy("http://proxy:port")

This sets your proxy server.

## set\_proxy\_authinfo

    $r->set_proxy_authinfo($user, $password)

If you have set a proxy and your proxy requires a username and password
you can set it with this.

## set\_type

    use HTTP::GHTTP qw/:methods/;
    my $r = HTTP::GHTTP->new();
    $r->set_uri('http://axkit.com/');
    $r->set_type(METHOD_HEAD);

This sets the request type. The request types themselves are constants
that are not exported by default. To export them, specify the `:methods`
option in the import list:

The available methods are:

- METHOD\_GET
- METHOD\_POST
- METHOD\_OPTIONS
- METHOD\_HEAD
- METHOD\_PUT
- METHOD\_DELETE
- METHOD\_TRACE
- METHOD\_CONNECT
- METHOD\_PROPFIND
- METHOD\_PROPPATCH
- METHOD\_MKCOL
- METHOD\_COPY
- METHOD\_MOVE
- METHOD\_LOCK
- METHOD\_UNLOCK

## set\_uri

    my $boolean = $r->set_uri($some_URI_string);

This sets the URI for the request.  It returns true if the URI was
properly set and undef otherwise.

# ASYNC OPERATION

It's possible to use an asynchronous mode of operation with [HTTP::GHTTP](https://metacpan.org/pod/HTTP::GHTTP). Here's
a brief example of how:

    my $r = HTTP::GHTTP->new("http://axkit.org/");
    $r->set_async;
    $r->set_chunksize(1);
    $r->prepare;

    my $status;
    while ($status = $r->process) {
        # do something
        # you can do $r->get_body in here if you want to
        # but it always returns the entire body.
    }

    die "An error occured" unless defined $status;

    print $r->get_body;

Doing timeouts is an exercise for the reader (hint: lookup select() in
perlfunc).

Note also that $sock above is an [IO::Handle](https://metacpan.org/pod/IO::Handle), not an [IO::Socket](https://metacpan.org/pod/IO::Socket), although
you can probably get away with re-blessing it. Also note that by calling
`$r->get_socket()` you load [IO::Handle](https://metacpan.org/pod/IO::Handle), which probably brings a lot of
code with it, thereby obliterating a lot of the use for [libghttp](http://ftp.gnome.org/pub/gnome/sources/libghttp). So
use at your own risk :-)

# AUTHOR

Matt Sergeant, <`matt@sergeant.org`>.

# CONTRIBUTORS

- Chase Whitener, <`capoeirab@cpan.org`>.

# COPYRIGHT AND LICENSE

Copyright (c) 2000 Matt Sergeant.  All rights reserved.

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}
    - Re-apply patch from dsteinwand
    - Require Alien::GHTTP to get around FreeBSD shared library problem
    - Up requirement to version 0.005 of Alien::GHTTP

1.080_004 2016-04-26
    - XS revert for testing issues
    - Added a few extra simplistic tests
    - Turned on Travis CI testing

1.080_003 2016-04-15
    - Fixed a minor POD issue
    - required 0.003 of the Alien::GHTTP module

1.080_002 2016-04-07
    - Fixed a missing dep
    - Fixed the date of libghttp's demise to 2002
    - Updated XS code to handle refcounts better (thanks, dsteinwand)
    - Use XSLoader instead of DynaLoader
    - Added XS_VERSION for test releases

1.080_001 2016-04-07
    - Made to use Dist::Zilla
    - Require Alien::GHTTP
    - Mark as deprecated
    - Cleaned up some documentation

1.07
    - Added support for clean() and NULL return values (david Ramboz)

1.06 Thu Jan 18 2000
    - Fixed "use of uninitialized value" messages
    - Minor tweaks

1.05 Wed Jan 17 2000
    - Added *much* better LIBS/INC code
    - Added get_headers() (works with libghttp 1.08 and higher)

1.04 Sat Jan 13 2000
    - Better INC and LIBS control.
    - Added g-request to MANIFEST

1.03 Wed Jan 10 2000
    - Use gnome-config to get INC and LIB from

1.02 Thu Nov 23 2000
    - Added asynchronous mode code and documentation

1.01 Wed Nov 22 2000
    - Bug fix in returned body string (could become corrupted with >1 request)
    - Added lots more ghttp methods (allows you to POST now and much more)
    - Documentation updates

1.0  Tue Nov 21 2000
    - original version


Makefile.PL
# This Makefile.PL for HTTP-GHTTP was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.38.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.006;
use ExtUtils::MakeMaker;

use Alien::GHTTP;

my %WriteMakefileArgs = (
  "ABSTRACT" => "(DEPRECATED) Perl interface to the gnome ghttp library",
  "AUTHOR" => "Matt Sergeant <matt\@sergeant.org>",
  "CONFIGURE_REQUIRES" => {
    "Alien::GHTTP" => "0.005",
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "HTTP-GHTTP",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "HTTP::GHTTP",
  "PREREQ_PM" => {
    "Alien::GHTTP" => "0.005",
    "Carp" => 0,
    "Exporter" => 0,
    "IO::Handle" => 0,
    "XSLoader" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "1.080_005",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

%WriteMakefileArgs = (
    %WriteMakefileArgs,
    LIBS => Alien::GHTTP->libs,
    CCFLAGS => Alien::GHTTP->cflags,
);

my %FallbackPrereqs = (
  "Alien::GHTTP" => "0.005",
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Test::More" => "0.88",
  "XSLoader" => 0,
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/HTTP/GHTTP.pm
package HTTP::GHTTP;

use strict;
use warnings;
use Alien::GHTTP;
use XSLoader ();
use base qw(Exporter);

our $VERSION = '1.080_005';
our $XS_VERSION = $VERSION;
$VERSION = eval $VERSION;

XSLoader::load('HTTP::GHTTP', $XS_VERSION);

our @EXPORT_OK = qw(
    get
    METHOD_GET
    METHOD_POST
    METHOD_OPTIONS
    METHOD_HEAD
    METHOD_PUT
    METHOD_DELETE
    METHOD_TRACE
    METHOD_CONNECT
    METHOD_PROPFIND
    METHOD_PROPPATCH
    METHOD_MKCOL
    METHOD_COPY
    METHOD_MOVE
    METHOD_LOCK
    METHOD_UNLOCK
);

our %EXPORT_TAGS = (
    methods => [
        qw(
            METHOD_GET
            METHOD_POST
            METHOD_OPTIONS
            METHOD_HEAD
            METHOD_PUT
            METHOD_DELETE
            METHOD_TRACE
            METHOD_CONNECT
            METHOD_PROPFIND
            METHOD_PROPPATCH
            METHOD_MKCOL
            METHOD_COPY
            METHOD_MOVE
            METHOD_LOCK
            METHOD_UNLOCK
            )
    ],
);

sub new {
    my $class = shift;
    my $r     = $class->_new();
    bless $r, $class;
    if (@_) {
        my $uri = shift;
        die "Blank uri not supported" unless length($uri);
        $r->set_uri($uri);
        while (@_) {
            my ($header, $value) = splice(@_, 0, 2);
            $r->set_header($header, $value);
        }
    }
    return $r;
}

sub get {
    die "get() requires a URI as a parameter" unless @_;
    my $r = __PACKAGE__->new(@_);
    $r->process_request;
    return $r->get_body();
}

sub get_socket {
    my $self = shift;
    require IO::Handle;
    my $sockno = $self->_get_socket();
    my $sock = IO::Handle->new_from_fd($sockno, "r")
        || die "Cannot open socket: $!";
    return $sock;
}

1;
__END__

=head1 NAME

HTTP::GHTTP - (DEPRECATED) Perl interface to the gnome ghttp library

=head1 SYNOPSIS

  use strict;
  use warnings;
  use HTTP::GHTTP;

  my $r = HTTP::GHTTP->new();
  $r->set_uri("http://axkit.org/");
  $r->process_request;
  print $r->get_body;

=head1 DESCRIPTION

The GNOME L<libghttp|http://ftp.gnome.org/pub/gnome/sources/libghttp> project
has been dead since 2002.  Thus, this library is deprecated.

B<Do not use this module!>

This is a fairly low level interface to L<libghttp|http://ftp.gnome.org/pub/gnome/sources/libghttp>,
which allows you to process HTTP requests to HTTP servers. There also
exists a slightly higher level interface - a simple get() function
which takes a URI as a parameter. This is not exported by default, you
have to ask for it explicitly.

=head1 FUNCTIONS

The L<HTTP::GHTTP> module makes the following standalone functions available.

=head2 get

  get($uri, [%headers])

This does everything automatically for you, retrieving the body at
the remote URI. Optionally pass in headers.

=head1 METHODS

The L<HTTP::GHTTP> module makes the following methods available.

=head2 new

  my $r = HTTP::GHTTP->new("http://axkit.com/", Connection => "close");

Constructor function - creates a new GHTTP object. If supplied a URI
it will automatically call set_uri for you. If you also supply a list
of key/value pairs it will set those as headers:

=head2 current_status

  $r->current_status()

This is only useful in async mode. It returns 3 values: The current
processing stage (0 = none, 1 = request, 2 = response headers,
3 = response), the number of bytes read, and the number of bytes total.

=head2 get_body

  my $rresponse = $r->get_body();

This gets the body of the response

=head2 get_error

  my $error = $r->get_error();

If the response failed for some reason, this returns a textual error.

=head2 get_header

  my $value = $r->get_header($header);

This gets the value of an incoming HTTP response header.

=head2 get_headers

  my @headers = $r->get_headers;
  print join("\n",
    map { "$_: " . $r->get_header($_) } @headers), "\n\n";

Returns a list of all the response header names in the order they
came back. This method is only available in L<libghttp|http://ftp.gnome.org/pub/gnome/sources/libghttp> 1.08 and later -
C<perl Makefile.PL> should have reported whether it found it or not.

=head2 get_socket

  my $sock_h = $r->get_socket();

Returns an L<IO::Handle> object that is the currently in progress socket.
Useful only when doing async downloads. There appears to be some corruption
when using the socket to retrieve file contents on more recent L<libghttp|http://ftp.gnome.org/pub/gnome/sources/libghttp>.

=head2 get_status

  $r->get_status()

This returns 2 values, a status code (numeric) and a status reason
phrase. A simple example of the return values would be (200, "OK").

=head2 prepare

  $r->prepare()

This is a low level interface useful only when doing async downloads.
See L<ASYNC OPERATION>.

=head2 process

  $r->process()

This is a low level interface useful only when doing async downloads.
See L<ASYNC OPERATION>.

process returns C<undef> for error, 1 for "in progress", and zero for
"complete".

=head2 process_request

  $r->process_request()

This sends the actual request to the server

=head2 set_async

  $r->set_async()

This turns async mode on. There is no corresponding unset function.

=head2 set_authinfo

  $r->set_authinfo($user, $password)

This sets an outgoing username and password for simple HTTP
authentication.

=head2 set_body

  $r->set_body($body);

This sets the body of a request, useful in C<POST> and some of the DAV
request types.

=head2 set_chunksize

  $r->set_chunksize($bytes)

Sets the download (and upload) chunk size in bytes for use in async
mode. This may be a useful value to set for slow modems, or perhaps
for a download progress bar, or just to allow periodic writes.

=head2 set_header

  $r->set_header($header_name, $header_value);
  $r->set_header('Connection', 'Close');

This sets an outgoing HTTP request header.

=head2 set_proxy

  $r->set_proxy("http://proxy:port")

This sets your proxy server.

=head2 set_proxy_authinfo

  $r->set_proxy_authinfo($user, $password)

If you have set a proxy and your proxy requires a username and password
you can set it with this.

=head2 set_type

  use HTTP::GHTTP qw/:methods/;
  my $r = HTTP::GHTTP->new();
  $r->set_uri('http://axkit.com/');
  $r->set_type(METHOD_HEAD);

This sets the request type. The request types themselves are constants
that are not exported by default. To export them, specify the C<:methods>
option in the import list:


The available methods are:

=over

=item *

METHOD_GET

=item *

METHOD_POST

=item *

METHOD_OPTIONS

=item *

METHOD_HEAD

=item *

METHOD_PUT

=item *

METHOD_DELETE

=item *

METHOD_TRACE

=item *

METHOD_CONNECT

=item *

METHOD_PROPFIND

=item *

METHOD_PROPPATCH

=item *

METHOD_MKCOL

=item *

METHOD_COPY

=item *

METHOD_MOVE

=item *

METHOD_LOCK

=item *

METHOD_UNLOCK

=back

=head2 set_uri

  my $boolean = $r->set_uri($some_URI_string);

This sets the URI for the request.  It returns true if the URI was
properly set and undef otherwise.

=head1 ASYNC OPERATION

It's possible to use an asynchronous mode of operation with L<HTTP::GHTTP>. Here's
a brief example of how:

    my $r = HTTP::GHTTP->new("http://axkit.org/");
    $r->set_async;
    $r->set_chunksize(1);
    $r->prepare;

    my $status;
    while ($status = $r->process) {
        # do something
        # you can do $r->get_body in here if you want to
        # but it always returns the entire body.
    }

    die "An error occured" unless defined $status;

    print $r->get_body;

Doing timeouts is an exercise for the reader (hint: lookup select() in
perlfunc).

Note also that $sock above is an L<IO::Handle>, not an L<IO::Socket>, although
you can probably get away with re-blessing it. Also note that by calling
C<< $r->get_socket() >> you load L<IO::Handle>, which probably brings a lot of
code with it, thereby obliterating a lot of the use for L<libghttp|http://ftp.gnome.org/pub/gnome/sources/libghttp>. So
use at your own risk :-)

=head1 AUTHOR

Matt Sergeant, <F<matt@sergeant.org>>.

=head1 CONTRIBUTORS

=over

=item *

Chase Whitener, <F<capoeirab@cpan.org>>.

=back

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2000 Matt Sergeant.  All rights reserved.

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
